File: src/main/java/org/nanoboot/spriteutils/core/SpriteSheetRow.java
///////////////////////////////////////////////////////////////////////////////////////////////
// sprite-utils: Tool used to work with sprites
// Copyright (C) 2024 the original author or authors.
//
// This program is free software: you can redistribute it and/or
// modify it under the terms of the GNU General Public License
// as published by the Free Software Foundation, either version 3
// of the License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see 
// <https://www.gnu.org/licenses/> or write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
///////////////////////////////////////////////////////////////////////////////////////////////
package org.nanoboot.spriteutils.core;

import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import lombok.Data;
import lombok.ToString;

/**
 * Represents a row in a sprite sheet.
 * This class is responsible for parsing and serializing sprite sheet rows from/to CSV format.
 * 
 * @author robertvokac
 */
@Data
@ToString
public class SpriteSheetRow {
    private static final String DELIMITER = ";";
    private static final String ID_DELIMITER = "__";
    private static final int MINIMUM_COLUMNS = 10;
    String file;
    String group;
    int numberInGroup;
    int row;
    int column;
    int x;
    int y;
    int width;
    int height;
    String notes;
    String tags;
    int numberPerSheet;
   /**
     * Constructor to create a SpriteSheetRow from a CSV line.
     * 
     * @param csvLine the CSV line containing the sprite sheet row data
      @throws IllegalArgumentException if the CSV line does not contain enough columns or contains invalid data
     */
    public SpriteSheetRow(String csvLine) {
        String[] csvColumns = csvLine.split(DELIMITER);
        if (csvColumns.length < MINIMUM_COLUMNS) {
            throw new IllegalArgumentException("CSV line does not contain enough columns: " + csvLine);
        }
        try {
            int i = 0;
            file = csvColumns[i++];
            group = csvColumns[i++];
            numberInGroup = Integer.parseInt(csvColumns[i++]);
            row = Integer.parseInt(csvColumns[i++]);
            column = Integer.parseInt(csvColumns[i++]);
            x = parseOptionalInt(csvColumns[i++]).orElse(-1);
            y = Integer.parseInt(csvColumns[i++]);
            width = Integer.parseInt(csvColumns[i++]);
            var column_ = csvColumns[i++];
            height = (column_.length() == 0 || column_.equals("0")) ? 0 :Integer.parseInt(column_);
            notes = i < csvColumns.length ? csvColumns[i++] : "";
            tags = i < csvColumns.length ? csvColumns[i++] : "";
            numberPerSheet = i < csvColumns.length ? Integer.parseInt(csvColumns[i]) : 0;

            // Adjust height if the column is greater than 1
            if (column > 1) {
                height = -height;
            }
        } catch (NumberFormatException e) {
            throw new IllegalArgumentException("CSV line contains invalid number format: " + csvLine, e);
        }
    }
    /**
     * Creates an ID for the sprite sheet row.
     * 
     * @return the generated ID
     */
    public String createId() {
        return String.join(ID_DELIMITER, file, group, String.valueOf(numberInGroup));
    }
    /**
     * Converts the sprite sheet row to a CSV line.
     * 
     * @return the CSV line representation of the sprite sheet row
     */
    public String toCsvLine() {
        List<String> fields = Arrays.asList(
            file, group, String.valueOf(numberInGroup), String.valueOf(row),
            String.valueOf(column), String.valueOf(x), String.valueOf(y),
            String.valueOf(width), String.valueOf(height), notes, tags,
            String.valueOf(numberPerSheet)
        );
        return String.join(DELIMITER, fields);
    }
    /**
     * Parses an optional integer from a string.
     * 
     * @param value the string to parse
     * @return an Optional containing the parsed integer, or empty if the string is blank
     */
    private Optional<Integer> parseOptionalInt(String value) {
        try {
            return value.isBlank() ? Optional.empty() : Optional.of(Integer.parseInt(value));
        } catch (NumberFormatException e) {
            throw new IllegalArgumentException("Invalid integer format: " + value, e);
        }
    }
}

File: src/SpriteSheetRow.cpp
/*
 * MIT License
 * Copyright (c) 2024-2025 Robert Vokac
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */


#include "SpriteSheetRow.h"
#include <sstream>
#include <stdexcept>

#include "Utils.h"

std::optional<int> SpriteSheetRow::parseOptionalInt(const std::string& value) {
    if (value.empty()) {
        return std::nullopt;
    }
    try {
        return std::stoi(value);
    } catch (...) {
        throw std::invalid_argument("Invalid integer format: " + value);
    }
}

// --- constructor from CSV line ---

SpriteSheetRow::SpriteSheetRow(const std::string& csvLine)
{
    auto cols = Utils::split(csvLine, DELIMITER);

    if (cols.size() < MINIMUM_COLUMNS) {
        throw std::invalid_argument("CSV line does not contain enough columns: " + csvLine);
    }

    try {
        int i = 0;
        file = cols[i++];
        group = cols[i++];
        numberInGroup = std::stoi(cols[i++]);
        row = std::stoi(cols[i++]);
        column = std::stoi(cols[i++]);
        x = parseOptionalInt(cols[i++]).value_or(-1);
        y = std::stoi(cols[i++]);
        width = std::stoi(cols[i++]);

        std::string hstr = cols[i++];
        height = (hstr.empty() || hstr == "0") ? 0 : std::stoi(hstr);

        notes = i < cols.size() ? cols[i++] : "";
        tags = i < cols.size() ? cols[i++] : "";
        numberPerSheet = i < cols.size() ? std::stoi(cols[i]) : 0;

        if (column > 1) {
            height = -height;
        }
    }
    catch (...) {
        throw std::invalid_argument("CSV line contains invalid number format: " + csvLine);
    }
}

// --- createId ---
std::string SpriteSheetRow::createId() const {
    return file + ID_DELIMITER + group + ID_DELIMITER + std::to_string(numberInGroup);
}

// --- serialize back to CSV ---
std::string SpriteSheetRow::toCsvLine() const {
    std::ostringstream oss;
    oss << file << DELIMITER
        << group << DELIMITER
        << numberInGroup << DELIMITER
        << row << DELIMITER
        << column << DELIMITER
        << x << DELIMITER
        << y << DELIMITER
        << width << DELIMITER
        << height << DELIMITER
        << notes << DELIMITER
        << tags << DELIMITER
        << numberPerSheet;

    return oss.str();
}